design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,233 @@
+
+However, we may have a limit on the size of the URL and we will have an issue  if the certificate will grow in the future. To overcome this, we can split the  request. The first one will create the file. The following requests will populate it with the cert content. Then the last request will verify that the whole content has passed (with one of the methods described above). We can decide on the number  of requests dynamically according to the size of the cert.
+
+We have a security concern with this option. While the traffic is fully encrypted when using https, including the URL, [it's still possible that the URL will be visible in certain locations](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), while the request travels to its destination. It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.
+
+Thus, we will not use this option due to security concerns.
+
+#### Option 3 - Client sets its State once cert is injected
+This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code).
+
+In this case, we respond with 202 Accepted to the "inject-client-cert" request and start injecting the cert in a separate thread. The server then polls on the state of the authn-client container and waits until its State changes to some  agreed State (e.g "CertInjected"). This state will be set by the authn-client once it sees that the cert is installed.
+
+This option is not fully backwards-compatible and is quite complex so I wouldn't suggest to implement it.
+
+#### Option 4 - Send cert as part of bash script
+
+The issue with the STDIN that is not closing _may_ be an issue of the `tar` command that waits for more input and doesn't close the connection. Another option that we have is to change the command from `tar` to `sh` and send a bash script in the STDIN.
+The script can be something like:
+
+```
+#!/bin/sh
+
+cat > /etc/conjur/ssl/client.pem <<EOF
+#{cert-content}
+EOF
+
+chmod 644 /etc/conjur/ssl/client.pem
+exit
+```
+
+This approach has the following benefits:
+1. Short URL. Certificate is in the script. (avoid issues raised from option 2)
+1. Deterministic operation. You know exactly when it is done, because the connection is closed once it reaches exit.
+1. Custom error handling. You can get any error you wish for, handle it properly, and send it back.
+1. Versatility. Script can be customized later if needed to handle more issues in this area.
+1. Simple implementation
+
+This solution is implemented in [this PR](https://github.com/cyberark/conjur/pull/1789).
```

Comment:
@rafis3 @sashaCher do you see any security concern with option 4?

Reply:
I don't see any security concerns. Does it technically work better than using the old tar method?

Reply:
what do you mean by "better"? It does work for both successful and unsuccessful injections. You can see the logs i shared in [this PR](https://github.com/cyberark/conjur/pull/1789)'s description.